contracts/filters/MerkleCollateralFilter.sol

```diff
@@ -0,0 +1,188 @@
+     * @return merkleProof Merkle proof
+     */
+    function _merkleProof(
+        bytes calldata context,
+        uint256 nodeCount
+    ) internal pure returns (bytes32[] memory merkleProof) {
+        /* Reduce number of merkle nodes by 1 if last 32 bytes are empty */
+        if (bytes32(context[context.length - 32:]) == bytes32(0)) nodeCount -= 1;
+
+        /* Instantiate merkle proof array */
+        merkleProof = new bytes32[](nodeCount);
+
+        /* Declare offset */
+        uint256 offset;
+
+        /* Populate merkle proof array */
+        for (uint256 i; i < nodeCount; i++) {
+            /* Set node */
+            merkleProof[i] = bytes32(context[offset:]);
+
+            /* Compute next offset */
+            offset += 32;
+        }
+    }
+
+    /**************************************************************************/
+    /* Getters */
+    /**************************************************************************/
+
+    /**
+     * @inheritdoc CollateralFilter
+     */
+    function COLLATERAL_FILTER_NAME() external pure override returns (string memory) {
+        return "MerkleCollateralFilter";
+    }
+
+    /**
+     * @inheritdoc CollateralFilter
+     */
+    function COLLATERAL_FILTER_VERSION() external pure override returns (string memory) {
+        return "1.0";
+    }
+
+    /**
+     * @notice Get collateral token
+     * @return Collateral token contract
+     */
+    function collateralToken() external view override returns (address) {
+        return _token;
+    }
+
+    /**
+     * @notice Get merkle root
+     * @return Merkle root
+     */
+    function merkleRoot() external view returns (bytes32) {
+        return _root;
+    }
+
+    /**
+     * @notice Get metadata URI
+     * @return Metadata URI
+     */
+    function metadataURI() external view returns (string memory) {
+        return _metadataURI;
+    }
+
+    /**************************************************************************/
+    /* Implementation */
+    /**************************************************************************/
+
+    /**
+     * @inheritdoc CollateralFilter
+     */
+    function _collateralSupported(
+        address token,
+        uint256 tokenId,
+        uint256 index,
+        bytes calldata context
+    ) internal view override returns (bool) {
+        /* Validate token supported */
+        if (token != _token) return false;
+
+        /* Declare node count */
+        uint256 nodeCount = _nodeCount;
+
+        /* Compute proof length */
+        uint256 proofLength = nodeCount * 32;
+
+        /* Compute start and end index */
+        uint256 startIndex = index * proofLength;
+        uint256 endIndex = startIndex + proofLength;
+
+        /* Validate context length */
+        if (context.length < endIndex) revert InvalidContext();
+
+        /* Compute proof from context */
+        bytes32[] memory proof = _merkleProof(context[startIndex:endIndex], nodeCount);
```

Comment:
I think we can collapse all of this into
```
bytes calldata proofData = context[index * _nodeCount * 32:(index + 1) * _nodeCount * 32];

if (proofData.length != _nodeCount * 32) revert InvalidContext();

bytes32 leaf = ...;

return MerkleProof.verify(_extractProof(proofData), _root, leaf);
```

Reply:
We can just keep `_proofLength` instead of `_nodeCount` to remove the * 32 everywhere